Find route around barriers

prupani-7 opened the pull request:
## Description
PR to add a new Kotlin sample "Find route around barriers" in Routes and Logistics category. This sample finds a route that reaches all stops without crossing any barriers.

## Links and Data
Sample Epic: runtime/kotlin/issues/1695 

## What To Review

-  Review the code to make sure it is easy to follow like other samples on Android
- `README.md` and `README.metadata.json` files


<!-- OPTIONAL
## To Discuss
-->

<!-- OPTIONAL
## Screenshots
-->

shubham7109 commented:
Hopefully, you don't mind but I've updated the dialog to use view binding instead of the `findViewById<>`. And updated the switches to persist the state of the toggle to match the iOS sample.

shubham7109 reviewed (changes requested):
Really impressive work with this sample @prupani-7. This sample is a great app to showcase the capabilities of the Maps SDK when many aspects of the API are interacting together in relation to the Android views and the MapView. I just have a couple of comments, but really well-done sample here.

prupani-7 commented:
> Hopefully, you don't mind but I've updated the dialog to use view binding instead of the `findViewById<>`. And updated the switches to persist the state of the toggle to match the iOS sample.

Thanks, yes it is better to use view binding and also have all switches initialized at the start. 
I refactored the switches code, to remove redundancy while checking `findBestSequenceSwitch.isChecked`. It was done twice. Also set the `isChecked = false` on `firstStopSwitch` and `lastStopSwitch` when `findBestSequenceSwitch` is unchecked.

prupani-7 commented on find-route-around-barriers/src/main/java/com/esri/arcgismaps/sample/findroutearoundbarriers/MainActivity.kt:
```diff
@@ -0,0 +1,476 @@
+        }
+
+        // solve the route task
+        val results = routeParameters?.let { routeTask?.solveRoute(it) }
```
renamed to routeResults

shubham7109 reviewed (approved):
Changes look good 👍

prupani-7 commented on find-route-around-barriers/src/main/java/com/esri/arcgismaps/sample/findroutearoundbarriers/MainActivity.kt:
```diff
@@ -0,0 +1,468 @@
+        routeOverlay.graphics.clear()
+        // clear the directions list from the directions list view, if they exist
+        directionsList.clear()
+
```
Sounds good.

fpearson1111 commented on find-route-around-barriers/src/main/java/com/esri/arcgismaps/sample/findroutearoundbarriers/MainActivity.kt:
```diff
@@ -0,0 +1,468 @@
+                        }
+                        // set the viewpoint to the selected maneuver
+                        val geometry = directionsList[position].geometry
+                        geometry?.let { Viewpoint(it.extent, 20.0) }
+                            ?.let { mapView.setViewpoint(it) }
```
I think it's better to just use one `let` here because the result of the first `let` will never be null, only if the geometry is null.
```suggestion
                        geometry?.let { mapView.setViewpoint(Viewpoint(it.extent, 20.0)) }
```

fpearson1111 commented on find-route-around-barriers/src/main/java/com/esri/arcgismaps/sample/findroutearoundbarriers/MainActivity.kt:
```diff
@@ -316,21 +313,24 @@ class MainActivity : AppCompatActivity() {
-        val routeResults = routeParameters?.let { routeTask?.solveRoute(it) }
+        val routeResults = routeParameters.let { routeTask.solveRoute(it) }
 
         if (routeResults != null) {
```
Is this check still needed?

fpearson1111 reviewed (commented):
Sorry for the long delay in completing the first round of this review. I've gone through the full code now and it looks good except just a few comments.

Although I also tried the sample and for me the directions list isn't displaying anything. Can you try it out again on your device?

prupani-7 commented:
Thanks @fpearson1111 no worries!
yes I am able to see the directions list on my device
<img width="448" alt="image" src="https://user-images.githubusercontent.com/44178002/219116743-f38353f6-69ec-48c5-9634-7a2a5997e12b.png">
I tap on the map twice, that creates 2 points and displays a route between them and the directions too

fpearson1111 reviewed (approved):
I think the reason I wasn't seeing any directions is that my phone is in dark mode. When I switch it to light mode the instructions appear. I attached some screenshots:
<img src=https://user-images.githubusercontent.com/90323222/219363655-06b14d71-281e-4921-a8c6-f2eed7ee224f.jpg width=50% />
<img src=https://user-images.githubusercontent.com/90323222/219363687-9c1b0710-a1ca-4d99-aba8-8d43695955ae.jpg width=50% />

I think this sample should be updated so that the instructions still display in dark mode but I will approve anyway because if the fix is easy then I am happy for it to be merged without re-approval. Nice work :)

prupani-7 commented:
I think that is something, we will need to update on the samples, which display directions list. There are 2 more samples, which will need an update I think, which displays the directions text. I will create a follow up issue, because I am not sure, how much the scope of work will be at this moment. 

The three samples which will need an update will be 
`find route around barriers`
`navigate route`
`find-route`

prupani-7 commented:
Thanks for the feedback @fpearson1111 always helpful!

shubham7109 commented on find-route-around-barriers/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,11 @@
+<resources>
+
+    <!-- Base application theme. -->
+    <style name="AppTheme" parent="Theme.MaterialComponents.DayNight.DarkActionBar" >
```
```suggestion
    <style name="AppTheme" parent="Theme.AppCompat.Light.DarkActionBar" >
```
For now, samples should be set to be in light theme, we can take on supporting dark theme in a future issue.

shubham7109 commented on find-route-around-barriers/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,11 @@
+<resources>
+
+    <!-- Base application theme. -->
+    <style name="AppTheme" parent="Theme.MaterialComponents.DayNight.DarkActionBar" >
```
```suggestion
    <style name="AppTheme" parent="Theme.MaterialComponents.Light.DarkActionBar" >
```
For now, samples should be set to be in light theme, we can take on supporting dark theme in a future issue.

prupani-7 replied:
Thanks @shubham7109 makes sense!
